task/clamav-scan/0.1/clamav-scan.yaml

```diff
@@ -45,6 +45,11 @@ spec:
           memory: 512Mi
           cpu: 10m
       script: |
+        #!/usr/bin/env bash
+        set -euo pipefail
+        . /utils.sh
+        trap 'handle_error $(results.TEST_OUTPUT.path)' EXIT
```

Comment:
I realized, we cannot do this in every step, it would emit result multiple times, that might be confusing.

We may need to have different handlers
handle_error only for steps not generating TEST_OUTPUT, it will only print error and exit 0
handle_error_with_test_output, which is current implementation of handle_error

Reply:
![image](https://github.com/redhat-appstudio/build-definitions/assets/20207783/6378bcf7-e203-448f-9616-27c74a094762)
on the other hands it doesn't look bad, but still if task is not supposed to generate TEST_OUTPUT

Reply:
If there is an failure in a step of a tekton task, won't the steps stop immediately? Will the next steps of this task be still run?

Reply:
It is interesting:
When there is no `trap` in tasks, it runs as below: 
```
STEP-STEP1

/tekton/scripts/script-0-2scgd: line 3: undefined1: unbound variable
STEP-STEP2

2023/10/09 08:25:30 Skipping step because a previous step failed
```
But if we have `trap` in steps, it runs as below, I think the `trap` makes the step finally look succeed, so that we get multiple `TEST_OUTPUT`:
```
STEP-STEP1

/tekton/scripts/script-0-c6ws4: line 5: undefined1: unbound variable
{"result":"ERROR","timestamp":"1696840096","note":"Unexpected error: Script errored at command: echo $undefined1.","namespace":"default","successes":0,"failures":0,"warnings":0}
STEP-STEP2

/tekton/scripts/script-1-66n4h: line 5: undefined2: unbound variable
{"result":"ERROR","timestamp":"1696840097","note":"Unexpected error: Script errored at command: echo $undefined2.","namespace":"default","successes":0,"failures":0,"warnings":0}
```

Reply:
@MartinBasti @dirgim I am thinking to add the real failure to `note` of `TEST_OUTPUT`, maybe we can add some code before [this line](https://github.com/redhat-appstudio/hacbs-test/blob/main/test/utils.sh#L162): Once the `TEST_OUTPUT` exists, we can only print some message to tell users the task experienced error in the previous step but skip generating `TEST_OUTPUT` again?

Reply:
> If there is an failure in a step of a tekton task, won't the steps stop immediately?


But that's whole purpose of why we are adding TRAP, check tasks cannot fail according ADR, exit code must be 0

Reply:
I mean, if one step fails with unexpected shell error, then generate `TEST_OUTPUT` with the failing shell command in `note`, and the next step exit with `0` but won't generate `TEST_OUTPUT` again.

Reply:
My meaning is to add code to `handle_error()` like [this](https://github.com/redhat-appstudio/hacbs-test/pull/158/files#diff-b17e8443b3f0dd1a2f1701af585a3a6222f2631b717c934c6fede3c8b16837ecR173-R178), t hen we still call `trap 'handle_error $(results.TEST_OUTPUT.path)' EXIT` in the steps of tasks, then if one step fails with unexpected error, the failing command will be added to the `note` of `TEST_OUTPUT`, then the followup steps won't generate `TEST_OUTPUT` again, but it prints log `"Task experienced unexpected error in previous step. For details, check Tekton task result and log.` in console and exit with `0`

Reply:
To me it looks much cleaner to only return TEST_OUTPUT from steps which are supposed to return it.
We can make it conditional and return TEST_OUTPUT only when `handle_error` has path specified

Reply:
> To me it looks much cleaner to only return TEST_OUTPUT from steps which are supposed to return it. I.e. don't make it conditional, write it explicitly in code where output is returned. We can make it conditional and return TEST_OUTPUT only when `handle_error` has path specified

Ok, let me update https://github.com/redhat-appstudio/hacbs-test/pull/158.

Reply:
https://github.com/redhat-appstudio/hacbs-test/pull/158 is updated

Reply:
Updated to use the new functions in the latest hacbs-test:v1.1.6.